### DISCUSSION TITLE
Meetings

### DISCUSSION THREAD

**Discussion by hojelse** [git-truck/git-truck#172] [General]
Meetings.

Previous meetings:
- Mon 17 Jan 2022 https://docs.google.com/document/d/1rinUPIsP4oTtAjXZoUfV4P6p3EChBJwVl21mxU9LRUU/edit?usp=sharing

**Comment by tjomson** [comment]
Tue 08 February 2022
Agenda
- List of test repositories
- Project agreement has been approved :))
- Regular meetings - Tuesday is bad
- https://github.com/git-visual/git-visual/
  - https://github.com/orgs/git-visual/projects/1
- We have studied .git
  -  Understanding of the .git folder
    - Commit objects
      - The history is based on the commits. It is a linked list. Points to the previous commit
    - Tree objects
      - Repo structure
    - Blob objects
      - Compressed files (+ some)
    - Tag objects
      - Unsure if meaningful for structure
  - Parse hashes (sort of like https://www.npmjs.com/package/git-parse )
  - pydriller
- Web app vs. locally run (maybe both?) (we need to decide to choose stack)
  - npx git-truck
  - Concerns:
    - It runs at the speed of the local machine
  - Robust prototype
    - Good layout
    - Good version control analysis
    - Local tool: No need to download anything, everything is already there in the folder
    - Full historical analysis or only current revision
    - git blame can gives us a picture of the current state
    - Color intensity: 
      - File Dominance
      - Different developers (limited by amount of individually distinguishable colors)
      - Bar chart - who worked how much on what
      - Semantic zoom - Which part of the file belongs to whom? 
    - Docker?
    - Private repository: local, privacy concerns with public
    - Target audience, non-technical SM or PO: deployed web app
    - Easy of use: deployed web app

Understanding the evolution, the structure, and the way the developers contribute to a code base is challenging.
Lack of understanding and overview, results in systems gradually getting harder and more expensive to maintain.
We propose to build an easy-to-use tool, that combines structural and version control information in a system overview. We expect that, such a tool can increase understanding of the evolution and the way the developers contribute to the project, to enable taking preventative measures to code base degradation.

- Version control information
- Evolution
- Structure
- Contributions
- Overview
- Easy-to-use

1. Prototyping: What can we do
    - Extracting data from git into some useful data structure
2. Looking at relating work: What is a good visualization that gives structure, evolution and overview
    - Research and design
3. What’s the lowest hanging fruit: The simplest way to solve the problem


- Folder structure: Semantics encoded in it
  - Represent a folder (multimetric)
    - Size
    - Color: Authors, Hotspot
    - Trees are good to visualize horizontally, more real-estate
- Overview: 

Iteration #1
- Structure
- Number of commits pr. file


Polymetric: Show structure and other information at the same time

**Comment by tjomson** [comment]
Wed 23 Feb 2022

Agenda
- product demo
  - chart types
  - metrics with legend
  - treemap
- product coming up
  - user union in UI
  - handle merge (churn, for correctness)
- repo license
- define coldmap metric
- should heatmap have a cutoff point

Notes
- We could test treemap vs bubblemap on users to determine which should be deffault
- tjoms has 45% on bubblechart (which can't be right)
- small boxes disappear in tree map
- Heat map could show churn or most recently active
- Changing color on heat map could be nice
- Separate metric from coloring scheme
- Call metrics more meaningful things (fx. "recently changed" instead of "cold map") (heatmap could be "commit activity")
- Heat map could be limited to changes in x number of days (number could be changed in UI)
- We should be able to do queries on the data
- We need to test the product on users
- Collapsible sidebar
- highlight what we sre sure of
- Committers with same percentage should be repressented better
- This tool does not necessarily show the truth, it allows the user to make hypothesis
- Show percentage ownership of one person using saturation (50% ownership of 1 person, should color file 50%)
- To show truck factor, 100% ownership is what we want to know (or 95%)
- We are going file-based, not tech-basec
- License that allows everything but commercial use (or MIT if we want it permissive)
- Description why a metric is useful
- Should cold map show the oldest or newest file most powerfully?
- File name on hover
- Long file names overflow infobox
- Showing just folder names, not file names
- Show path for file in infobox
- Padding between files not necessary
- Treemap might be better for big projects, bubblemap for small projects

Goal
- In 7-14 days mircea should be able to run it locally

**Comment by hojelse** [comment]
Wed 02 March 2022

Notes
- To merge the parser and app, we could wrap it all in one project, or import one to the other
- To search in files we need to create indexes for things in the file to be performant
- Search will make more sense if you can search in file content
- Using a regex for search could be useful
- Search could also have check for case-sensitivity and/or matching full word
- Better naming for metrics
- Gradiant legend for heat map
- Using search on dominant author is smart because we can show two types of information at the same time
- Multiple searches would be overkill, we could call it "filter" instead (let's not change that until we have done user tests)
- Reviews are worth it, so it is better to take our time than rush
- Humans don't see small color differences well, so we could make some intervals instead of a gradiant
- On heatmap, make clear what the values used for the colors are
- On mouse over, show the value of a metric
- Tests are marked as binary files?
- We should decide what is useful, we can add more coloring options, but the default uses of each metric should be clear
- Adding icons for chart type are not necessary
- Churn can be defined as "number of commits per file"
- New metric based on ownership based on blame (changes per user for newest version of files)
- Slider where you can use "all history" or "most recent change"
- Churn is on an interval, and blame is at a specific time
- Make explicit how dominant authors are calculated (fx. based on churn)
- Many settings are not necessarily good, if only few configurations make sense
- We can be opinionated in how the user is "supposed" to use
- Missing description of what size means
- Add size metric (lines of code / number of bytes)
- Make regression testing on other repos that we clone a snapshot of instead of writing out own dummy repos
- We should focus on getting the user-union to work. It should also work on
- Parser and app should be able to read/write to the truckconfig
- This is meant to be a developer-tool
- Decide on whether parser runs before app or vice versa
- It is okay for developers, that a parser-rerun is necessary for ignoring files

Takeaways
- Focus on good code
- Prioritize the most important things to work on
- Focus is usability and showing more perspectives


For next time
- Run the tool on the student projects and ZeeGuu
- Then we discuss if it works on many projects
- We need the project to be easy to run for this

**Comment by tjomson** [comment]
Wed 09 march 2022

Notes
- Don't worry about merge commit testing
- It might/might not be faster in a docker container
- Mircea thinks that the most highlighted on cold map would make sense to be the newest change
- "Last changed" needs to be renamed
- Either use ISO date format or use month with letters
- It could still be useful to set time interval for most commits
- We need a definition of "to dominate"
- Maybe say "is the only author" instead of "dominates"
- We could show the author color on dominated files
- We can't see if one person dominates a lot
- "dominant author" could be renamed to "top contributer"
- "dominated files" could be renamed to "single author"
- Exclude muted colors from dominant author, so that the non-hihglighted files are the only ones that are muted
- Combining dominant author and dominated files into one by showing both
- Reserve colors for search result maybe
- Handle when files are empty without showing them as binary
- Show <1% instead of rounding down to 0%
- Having both truckconfig and unioning in UI is useful (add truckconfig info to readme)
- Shell scripts don't run everywhere
- We want to make it easy for our testers to just npm i -> npm run, not necessarily use shell scripts
- Being able to swap between branches could be cool
- Mircea wants to write a paper about how tools like gittruck are useful
- Maybe be able to set the intervals on the gradient manually on most commits metrics
- We need to detect colliding text fields
- Zooming or being able to show just one package would be useful for big repos. This could make breadcrumb to show what is shown
- We could show the file structure immediately, and then animate the hydration once that completes running

The plan forward
- Fixing bugs
- Renaming to make it more user friendly
- getting it ready to be tested (deploying on npm)
- Next week think about what we want testers to do
- Will it be better to run it from a node process instead of bash script?

Mircea was able to clone the project, and run the shell script

**Comment by tjomson** [comment]
Wed 16 march 2022

Notes
- For zooming, we could show the path in the upper left corner where the repo name is as *git-truck/parser* if parser is clicked
- Gradient for clicking a name could be cool, but not very intuitive
- Description would help on this, but it might be a filter instead
- Maybe add "highlight" as a dropdown to decide what should be highlighted
- Chart type gives us the layout, metric gives us the color. There could be added a filter, to only apply the color metric to specific things.
- Single author could be a filter
- Leave author gradient for a second iteration
- We could have a lot of settings, and some presets that set them to something we decide
- Removing the root circle to make circles bigger, could be cool, but is not necessary
- Time of last change should match tooltip should match the legend, so both should be relative
- Changing branch would be a useful feature
- Keep track of all user tests we make, so we can reason about our design process later on
- The focus is still to make it stable, fixing bugs, and testing it

**Reply by nimrossum** [reply]
What does

> Gradient for clicking a name could be cool, but not very intuitive

mean 🤔

**Comment by tjomson** [comment]
Wed 23 march 2022

Notes 
- Style path with blue underline to make it clear that it is clickable
- Mircea was able to use it with npx, but hit a problem with parser infinite loop (because the version on npm was old)
- We could check for new versions when it runs (tell user to npx git-truck@latest)
- Reset quotepath on termination
- Ask user if they want us to set quotepath
- Could we bundle our git version, so we don't depend on the user?
- Let's not worry too much about quotepath
- We should not update the gitignore for people with a truckconfig entry
- Next time we don't do demo, we have to make a plan for evalutation
- Think about the research question
- Like "if you give people a tool like this, what do they do with it? who uses it, how do they use it, what are they missing?"
- The research part is not to build a tool, but to get knowledge about how a tool like this is used
- Spam everyone with npx, and get them to answer a survey
- We may need a small pilot project to see if the instructions are clear
- Ask people "what did you learn?" "what would you pay/what do you need to pay for it?"
- Mircea will send us some example papers, to inspire us of how to do our study
- Important to prompt people to update if their version is out of date
- Button in UI to send us feedback, could use github issues
- Button to open issue, with issue template
- It is part of life, that the project is not entirely correct
- Info tooltip for menu items are not necessary right now, we will ask the users if the menu items are selfexplanatory enough
- We could batch together git commands to improve
- Metric description does not belong in legend
- It should be moved to below the metric-select
- Chart type is secondary to metric, so maybe it should be reordered
- Cold map should say "shows most recently changed" ish
- Top contributor text should explain that it is lines added/removed
- Short options with explanation are good for experienced and new users
- We could move some menu items to the right
- legend in bottom right, info in top right
- We could anchor legend to the bottom
- Button to hide all ui

Now:
- "Please update" notification
- Plan user tests

**Comment by tjomson** [comment]
Wed 30 march 2022

Notes
- Show how to update in UI, not on hover
- We let users decide where descriptive text should be
- Showing root in breadcrumb is not necessary
- Opening file does not work on linux
- For the study, if we ask people if we can send usage info about them, then it is fine
- All questions should be based on something we want to know
- If things go bad, we focus on that, if things go well, we ask about what they liked
- Interesting questiens:
  - Did you find an interesting usecase while using the tool?
  - Please put screenshots about useful configurations and such
- We need to make it such that people that find the tool useful, should have a way of showing us, like google form or something
- We should be able to disentangle the ideas from the tool, so someone can make a similar tool by just looking at our user test results
- More open question than "which chart was most useful?". Like "which charts are most useful?"
- Google form where users can put a picture and what they learned
- Who do we ask?
  - We ask the people that we think will be useful
  - Maybe ask students from second year project
  - Mircea can find some teachers
  - Try it on different kinds of projects
  - Sending it to companies
  - Let people figure out themselves whether it is useful
  - Ask about their role. Like student/teacher/developer/hobbyist
  - Context of when they use the tool, during meetings, at standups, etc.
 - Do usability study of whether people can use the tool from just the readme
 - Test usability of the whole process of acquiring a user
 - Small scale usability test first, then share it around
 - A research question: "for who and in which cases is the tool useful?"
 - If we can find the few thousand developers in the world that need this tool, then it is already amazing
 - Explain why the tool is called git-truck in readme
   - Either explain the convoluted story of the name, or just say "a truck load of git info"
 - Plan the easiest onboard process, and test it on a few people
 - Maybe present the project in a video
 - Screenshot of tool in readme
 - Command line options should be low in the readme
 - Figure out research questions and interview questions
 - Our tool is sort of like GitLens Pro, which they charge 9 dollars a month for
   - We could research what users use GitLens for
 - We could show i bit of the file when you are zoomed in very far
 - Maybe move the details panel to somewhere else, so that it does not seem like it is connected to the legend
 - Maybe add value to the legend arrow

**Comment by tjomson** [comment]
Wed 6 april 2022

Notes
- Controlled experiment if we know what we want to know
- Ease of use is not a research question, but it is important to us
- Don't spend too much time on the ease of use part, spend more time on the usefulness
- RQ2 is a good research question
- User's relation to project is important. Are they new to a project, do they work on it daily? How many people are working on it? Is it open source so we can look at it?
- RQ2 is generall good
- RQ3 is a bit vaguely formulated, and seems like a contiuation of RQ2
- We should not be held hostage by the project description we made at the start of the project
- Split the question about whether people want  to pay into "what do you need to use it monthly? weekly? daily? to be willing to pay?"
- Ask user if using the tool, made them want to something differently in their project
  - Did it make them want to investigate something?
  - Quantify impact of the tool (did the user want to check something out? plan a change? make a change right now?)
- Questions are in principle quite good
- Keep looking at other papers to see how they do user tests
- Balance between open and closed questions
- Have holiday if we need it to not burn out

**Comment by emiljapelt** [comment]
Wed 27 april 2022

Points
- Get DevOps & TID project groups to try the tool and fill out the survey.
- Where else can we get people to try the tool?
- Performance
- Branch/Tag selection in UI
- Open folder of repos
- Questionaire
- A lead for user testing

Notes:
- Most fun project Mircea has worked with on ITU :D
- Being able to select tags is good, because it encourages people to use tags
- Maybe not use npx, if it causes problems. We should expect people to be able to run two commands
- We should try to get this out to a lot of people (reddit maybe)
- Maybe we should have 2-factor authentication on npm before sending it out to a lot of people
- Not sending statistics automatically is a shame, but it means that we don't have to have a privacy policy
- We should set up a meetup
- Setting up our own meetup might not work, but we can go to some event
- Mircea has email blueprint to invite people to study
- We need to make it clear that people help science by participating
  - "it is research, and the goal is to understand evolution"
- Only have one way to do things in readme
- "if you already installed it, please run it, otherwise please run it"
- People trying out the tool without giving feedback, is not useful to us
- Otherwise we have to send automatic feedback
- Give people an indicator of how long they are supposed to wait for the analyzer
- Only ask people for version of npm, git, etc. if it does not work for them
- If we get more responses and people answer the same thing, we can remove that question
- People from facebook did need the readme, because the post had all the info
- code degredation questions are difficult to understand
- Our examples of code degredation is highly speculative, maybe remove that section
- Maybe both asking people about their role, and their day-to-day relation is redundant
- Maybe ask people to just try one project, and give us feedback on that
- Ask people how big the project is (number of files)
- Anchor people to just one project, and ask more about what they learned on that specific project

**Comment by tjomson** [comment]
Wed 04 may 2022

Agenda
- Finalizing survey
- How do we get people to answer?
   - React meetup maybe
- Report structure and where to start
- Product news
   - Union authors
   - Performance improvements
- Should we work on more features, or leave it?

No notes were taken, as we just went through the survey and refined it.

**Comment by tjomson** [comment]
Wed 11 may 2022

Agenda
- Title
- Merge introduction and problem
- Presenting solution
  - How much detail to describe features
- Relevance of "ease of use" for research
- Describe process vs outcome
- How to present multiple results in a good way

Notes
- More structured the better
- The reader needs to know what problem to 
- Look at the answers, and see what is most interesting
- Being able to say "top contributor this week"
- "future work" section should just be the top 3 directions to go
- Discussion section after the result
- Discussion in a separate section than results. Results should be as factual as possible
- Just 3 out of 13 people finding it useful is also a success
- If we get more responses between the handin and the presentation, we can talk about it at the exam
- If people did not put any responses, we must assume they have nothing to say.
- The possible uses that users suggest should be mentioned as possible directions
- It might be useful to familiarize yourself with a new project
- Main uses: first time seeing a project, and when doing big refactor
- We are a tool to view a whole project on one screen. Who is that for?
- Future work should propose the next 3 projects, not just more features
  - Improve tool for refactoring, improve for onboarding (see history of file/folder)
  - Word cloud generated from commit messages
  - See key words on files based on function names
  - Throw away frequent words (like "for", "if", "function")
  - TL;DRepo
  - Search within files
- We have an answer to research question "did it increase understanding...?", "what persona is it useful for...?"
- How might we design an easy to use repo visualization tool for developers?
  - We focus on students, but found that academics also found it useful
- Having "developers" in the problem would allow us to discover that it is also nice for academics
- The research question should be part of the introduction (not necessarily in the form of a question)
- Say that there might be tools, but nobody made a study about it, and there are studies, but they are about difficult to use tools.
- Leonnel Merino
  - use http://scg.unibe.ch/archive/papers/Meri18a.pdf as a source to conclude that people don't evaluate much
- A research question should stand out (bold text, or RQ next to it)
- Start with research question, and then describe why we needed to make a tool to do it, and then how we made the tool. How did we gather data (twitter, LinkedIn, Reddit), what did we find
- Subquestions for the research question. Each section in "results" should answer a question
- Things not answering research questions go in "discussion" section

**Comment by nimrossum** [comment]
Agenda
- How should we submit Link to GitHub repository in LearnIT - should we just upload a txt file with the link in it?
